pkg/routeagent_driver/handlers/healthchecker/healthchecker.go

```diff
@@ -0,0 +1,325 @@
+	}()
+
+	return nil
+}
+
+// TransitionToNonGateway is called once for each transition of the local node from Gateway to a non-Gateway.
+func (h *controller) TransitionToNonGateway() error {
+	h.Lock()
+	defer h.Unlock()
+
+	if h.config.HealthCheckerEnabled {
+		remoteEndpoints := h.State().GetRemoteEndpoints()
+
+		for i := range remoteEndpoints {
+			err := h.processEndpointCreatedOrUpdated(&remoteEndpoints[i])
+			if err != nil {
+				logger.Warningf("Error processing remote endpoint %s: %v", remoteEndpoints[i].Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// TransitionToGateway is called once for each transition of the local node from non-Gateway to a Gateway.
+func (h *controller) TransitionToGateway() error {
+	h.Lock()
+	defer h.Unlock()
+
+	if h.config.HealthCheckerEnabled {
+		for i := range h.pingers {
+			h.pingers[i].Stop()
+			delete(h.pingers, i)
+		}
+	}
+
+	h.syncRouteAgentStatus()
+
+	return nil
+}
+
+func (h *controller) GetNetworkPlugins() []string {
+	return []string{event.AnyNetworkPlugin}
+}
+
+func (h *controller) GetName() string {
+	return "routeAgent-health-checker"
+}
+
+func (h *controller) syncRouteAgentStatus() {
+	routeAgent := h.generateRouteAgentObject()
+	routeAgent.Status.RemoteEndpoints = []submarinerv1.RemoteEndpoint{}
+	remoteEndpoints := h.State().GetRemoteEndpoints()
+
+	for i := range remoteEndpoints {
+		pingerObject, found := h.pingers[remoteEndpoints[i].Spec.CableName]
+		if !found {
+			logger.Warningf("Pinger not found for %q", remoteEndpoints[i].Spec.CableName)
+			continue
+		}
+
+		latencyInfo := pingerObject.GetLatencyInfo()
+		var connectionStatus submarinerv1.ConnectionStatus
+		var remoteEndpoint submarinerv1.RemoteEndpoint
+		var statusMessage string
+
+		if !h.config.HealthCheckerEnabled {
+			connectionStatus = submarinerv1.ConnectionNone
+			statusMessage = "Health check is not enabled"
+		} else if h.State().IsOnGateway() {
+			connectionStatus = submarinerv1.ConnectionNone
+			statusMessage = "Health check is not performed on gateway nodes"
+		} else if latencyInfo != nil {
+			switch latencyInfo.ConnectionStatus {
+			case pinger.Connected:
+				connectionStatus = submarinerv1.Connected
+				statusMessage = ""
+			case pinger.ConnectionError:
+				connectionStatus = submarinerv1.ConnectionError
+				statusMessage = latencyInfo.ConnectionError
+			case pinger.ConnectionUnknown:
+				connectionStatus = submarinerv1.ConnectionError
+				statusMessage = latencyInfo.ConnectionError
+			}
+		} else {
+			connectionStatus = submarinerv1.Connecting
+			statusMessage = ""
+		}
+
+		remoteEndpoint = submarinerv1.RemoteEndpoint{
+			Status:        connectionStatus,
+			StatusMessage: statusMessage,
+			Spec:          remoteEndpoints[i].Spec,
+			LatencyRTT: &submarinerv1.LatencyRTTSpec{
+				Last:    latencyInfo.Spec.Last,
+				Min:     latencyInfo.Spec.Min,
+				Average: latencyInfo.Spec.Average,
+				Max:     latencyInfo.Spec.Max,
+				StdDev:  latencyInfo.Spec.StdDev,
```

Comment:
`latencyInfo` could be nil here. I'd suggest initializing the `RemoteEndpoint` above and filling in the fields appropriately:

```
remoteEndpoint := submarinerv1.RemoteEndpoint{
    Status: submarinerv1.Connecting,
    Spec:          remoteEndpoints[i].Spec,
}

if !h.config.HealthCheckerEnabled {
    remoteEndpoint.Status = submarinerv1.ConnectionNone
    remoteEndpoint.StatusMessage = "Health check is not enabled"
} else if h.State().IsOnGateway() {
    remoteEndpoint.Status = submarinerv1.ConnectionNone
    remoteEndpoint.StatusMessage = "Health check is not performed on gateway nodes"
} else if pingerObject, found := h.pingers[remoteEndpoints[i].Spec.CableName]; found {
     latencyInfo := pingerObject.GetLatencyInfo()
     if latencyInfo != nil {
         switch latencyInfo.ConnectionStatus {
             ....
	 }

         remoteEndpoint.LatencyRTT =  &submarinerv1.LatencyRTTSpec {
             ...
         }
     }
} 

routeAgent.Status.RemoteEndpoints = append(routeAgent.Status.RemoteEndpoints, remoteEndpoint)
```

Reply:
Refactored the code.